File: backend/app/schemas/characterization.py
"""
GreenSynth Analytics — Characterization & RawFile Pydantic Schemas

Validation schemas and technique-to-allowed-format mappings.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class TechniqueType(StrEnum):
    XRD = "XRD"
    UV_VIS = "UV_VIS"
    FTIR = "FTIR"
    SEM = "SEM"
    ELECTRICAL = "ELECTRICAL"


class CharacterizationStatus(StrEnum):
    UPLOADED = "UPLOADED"
    READY_FOR_ANALYSIS = "READY_FOR_ANALYSIS"
    PROCESSING = "PROCESSING"
    ANALYZED = "ANALYZED"
    ARCHIVED = "ARCHIVED"


# ── Allowed file extension mappings per technique ───────────
TECHNIQUE_ALLOWED_EXTENSIONS: dict[TechniqueType, set[str]] = {
    TechniqueType.XRD: {"csv", "txt", "xlsx", "json"},
    TechniqueType.UV_VIS: {"csv", "txt", "xlsx", "json"},
    TechniqueType.FTIR: {"csv", "txt", "xlsx", "json"},
    TechniqueType.ELECTRICAL: {"csv", "txt", "xlsx", "json"},
    TechniqueType.SEM: {"png", "jpg", "jpeg", "tiff", "tif", "pdf"},
}


# ── Characterization Schemas ────────────────────────────────

class CharacterizationBase(BaseModel):
    technique: TechniqueType
    characterization_date: datetime | None = None
    operator: str | None = Field(default=None, max_length=255)
    instrument_name: str | None = Field(default=None, max_length=255)
    instrument_model: str | None = Field(default=None, max_length=255)
    instrument_id: str | None = Field(default=None, max_length=128)
    notes: str | None = None


class CharacterizationCreate(CharacterizationBase):
    sample_id: UUID


class CharacterizationUpdate(BaseModel):
    characterization_date: datetime | None = None
    operator: str | None = Field(default=None, max_length=255)
    instrument_name: str | None = Field(default=None, max_length=255)
    instrument_model: str | None = Field(default=None, max_length=255)
    instrument_id: str | None = Field(default=None, max_length=128)
    notes: str | None = None
    status: CharacterizationStatus | None = None


class RawFileResponse(BaseModel):
    id: UUID
    characterization_id: UUID
    sample_id: UUID
    original_filename: str
    stored_filename: str
    file_extension: str
    mime_type: str | None
    file_size: int
    checksum: str
    storage_path: str
    uploaded_at: datetime
    uploaded_by: str | None
    status: str

    model_config = ConfigDict(from_attributes=True)


class CharacterizationResponse(CharacterizationBase):
    id: UUID
    sample_id: UUID
    status: CharacterizationStatus
    raw_files: list[RawFileResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/characterization.ts
/**
 * GreenSynth Analytics — Characterization & RawFile TypeScript Types (Phase 3)
 */

export type TechniqueType = 'XRD' | 'UV_VIS' | 'FTIR' | 'SEM' | 'ELECTRICAL'

export type CharacterizationStatus =
  | 'UPLOADED'
  | 'READY_FOR_ANALYSIS'
  | 'PROCESSING'
  | 'ANALYZED'
  | 'ARCHIVED'

export interface RawFile {
  id: string
  characterization_id: string
  sample_id: string
  original_filename: string
  stored_filename: string
  file_extension: string
  mime_type: string | null
  file_size: number
  checksum: string
  storage_path: string
  uploaded_at: string
  uploaded_by: string | null
  status: string
}

export interface Characterization {
  id: string
  sample_id: string
  technique: TechniqueType
  characterization_date: string | null
  operator: string | null
  instrument_name: string | null
  instrument_model: string | null
  instrument_id: string | null
  notes: string | null
  status: CharacterizationStatus
  raw_files: RawFile[]
  created_at: string
  updated_at: string
}

export interface CharacterizationCreate {
  sample_id: string
  technique: TechniqueType
  characterization_date?: string
  operator?: string
  instrument_name?: string
  instrument_model?: string
  instrument_id?: string
  notes?: string
}

export interface CharacterizationUpdate {
  characterization_date?: string
  operator?: string
  instrument_name?: string
  instrument_model?: string
  instrument_id?: string
  notes?: string
  status?: CharacterizationStatus
}

export const TECHNIQUE_ALLOWED_EXTENSIONS: Record<TechniqueType, string[]> = {
  XRD: ['csv', 'txt', 'xlsx', 'json'],
  UV_VIS: ['csv', 'txt', 'xlsx', 'json'],
  FTIR: ['csv', 'txt', 'xlsx', 'json'],
  ELECTRICAL: ['csv', 'txt', 'xlsx', 'json'],
  SEM: ['png', 'jpg', 'jpeg', 'tiff', 'tif', 'pdf'],
}
